src/constructors.jl

```diff
@@ -90,6 +90,24 @@ julia> b ≈ c
 true
 ```
 
+# Extended help
+
+Advanced users can customize the implementation of the `@avx`-annotated block
+using keyword arguments:
+
+```
+@avx inline=false unroll=2 tile=(2,2) body
+```
+
+where `body` is the code of the block (e.g., `for ... end`).
+
+`inline` is a Boolean that ?doesn't do anything?.
+`unroll` is an integer that specifies the loop unrolling factor.
+`tile` a tuple of integers, one per loop (so `(2,2)` would be appropriate
+for a `for m ∈ 1:size(A,1), n ∈ 1:size(B,2)` loop). ?Do these specify the number
```

Comment:
Currently, `tile` means to unroll 2 loops. So `(3,4)` would mean to unroll one loop by a factor of 3, and another by a factor of 4.

What do you think would be the best naming convention? Repeated in the comment following this one, I'd like to replace the `tile` keyword with letting `unroll` accept `tuple` arguments.

Then, `unroll` would refer to loop unrolling, regardless of the number of loops being unrolled (although for now, the only possibilities are 0, 1, and 2).
While a `tile` would be much larger, and refer to evaluating pieces of the underlying array at a time for better cache efficiency -- an optimization `LoopVectorization` does not do yet.

Reply:
I think you made the right decision.

I'm still a little confused about `inline`, in my tests it doesn't actually seem to change the generated code at all and https://github.com/chriselrod/LoopVectorization.jl/blob/6f299fabff64b756ff18538bf767cb85fed133e1/src/condense_loopset.jl#L402-L406 does not entirely resolve my confusion.

Reply:
What were your tests?

Hmm -- maybe I should go back to three-possibilities:
1. `@inline`
2. `@noinline`
3. Let some cost model decide.

Currently, `inline=false` let's Julia's cost modeling decide. Perhaps I should make it `2.` instead.
Feel free to add a `pushfirst!(q.args, Expr(:meta,:noinline))` to `__avx__!` (note the double-underscores).

`_avx_!` (single underscores) is automatically inlined. With `inline = true` (the default), `@avx` generates a call to `_avx_!`.
With `inline=false`, it calls `__avx__!` instead. That's so that `__avx__!` can then take the pointers of arrays and calculate some offsets within the same function, hopefully providing LLVM with a little more information, but I'm sure if it really helps.
`_avx_!` itself currently expects stridedpointers. I thought that was the easiest/more trait-like way of supporting a variety of array types: they just have to define an appropriate `stridedpointer` method.

Because LoopVectorization generates a lot of code through all the loop unrolling and branches to handle remainders, I figured it is unlikely that `__avx__!` would be inlined.

For example:
```julia
julia> using LoopVectorization, BenchmarkTools

julia> function mydot(a, b)
           s = zero(promote_type(eltype(a), eltype(b)))
           @inbounds @simd for i ∈ eachindex(a,b)
               s += a[i] * b[i]
           end
           s
       end
mydot (generic function with 1 method)

julia> function mydotavx(a, b)
           s = zero(promote_type(eltype(a), eltype(b)))
           @avx for i ∈ eachindex(a,b)
               s += a[i] * b[i]
           end
           s
       end
mydotavx (generic function with 1 method)

julia> function mydotavx_noinline(a, b)
           s = zero(promote_type(eltype(a), eltype(b)))
           @avx inline = false for i ∈ eachindex(a,b)
               s += a[i] * b[i]
           end
           s
       end
mydotavx_noinline (generic function with 1 method)

julia> a = rand(31); b = rand(31);

julia> @btime mydot($a, $b)
  15.786 ns (0 allocations: 0 bytes)
6.91914940245568

julia> @btime mydotavx($a, $b)
  6.513 ns (0 allocations: 0 bytes)
6.91914940245568

julia> @btime mydotavx_noinline($a, $b)
  6.838 ns (0 allocations: 0 bytes)
6.91914940245568

julia> M = K = N = 128
128

julia> A = rand(M, K); B = rand(K, N); C1 = Matrix{Float64}(undef, M, N); C2 = similar(C1);

julia> a = view(A, 1:31, 1);

julia> b = view(B, 1:31, 1);

julia> @btime mydot($a, $b)
  17.855 ns (0 allocations: 0 bytes)
7.1870499901207525

julia> @btime mydotavx($a, $b)
  6.540 ns (0 allocations: 0 bytes)
7.187049990120752

julia> @btime mydotavx_noinline($a, $b)
  12.832 ns (0 allocations: 0 bytes)
7.187049990120752

julia> a = view(A, 1:32, 1);

julia> b = view(B, 1:32, 1);

julia> @btime mydot($a, $b)
  4.884 ns (0 allocations: 0 bytes)
7.214710063797762

julia> @btime mydotavx($a, $b)
  6.076 ns (0 allocations: 0 bytes)
7.214710063797762

julia> @btime mydotavx_noinline($a, $b)
  13.052 ns (0 allocations: 0 bytes)
7.214710063797762

julia> a = view(A, 1:127, 1);

julia> b = view(B, 1:127, 1);

julia> @btime mydot($a, $b)
  28.757 ns (0 allocations: 0 bytes)
32.21951520775383

julia> @btime mydotavx($a, $b)
  12.314 ns (0 allocations: 0 bytes)
32.21951520775384

julia> @btime mydotavx_noinline($a, $b)
  18.822 ns (0 allocations: 0 bytes)
32.21951520775384

julia> a = view(A, :, 1);

julia> b = view(B, :, 1);

julia> @btime mydot($a, $b)
  8.329 ns (0 allocations: 0 bytes)
32.22429313653614

julia> @btime mydotavx($a, $b)
  11.668 ns (0 allocations: 0 bytes)
32.22429313653614

julia> @btime mydotavx_noinline($a, $b)
  14.647 ns (0 allocations: 0 bytes)
32.22429313653614

julia> a = rand(128); b = rand(128);

julia> @btime mydot($a, $b)
  8.350 ns (0 allocations: 0 bytes)
29.895720677897188

julia> @btime mydotavx($a, $b)
  8.760 ns (0 allocations: 0 bytes)
29.895720677897188

julia> @btime mydotavx_noinline($a, $b)
  9.596 ns (0 allocations: 0 bytes)
29.895720677897188
```
(Note that I'm being very generous to LoopVectorization by picking 31 as the length.)
I don't know why `inline = false` sometimes causes such a steep regression, nor why using `view`s do.

The margins in the [dot product benchmarks](https://chriselrod.github.io/LoopVectorization.jl/latest/examples/dot_product/) are very slim. That was the most important factor in deciding to make `inline=true` the default.

We can look at the native code to confirm that with `inline=false`, the loop was not inlined, instead calling `__avx__!`:
```asm
# julia> @code_native debuginfo=:none mydotavx_noinline(A, B)
        .text
        subq    $40, %rsp
        movq    %rdi, %r9
        movq    8(%rdi), %rax
        movq    %rax, %rcx
        sarq    $63, %rcx
        andnq   %rax, %rcx, %rax
        movq    8(%rsi), %rcx
        movq    %rcx, %rdx
        sarq    $63, %rdx
        andnq   %rcx, %rdx, %rcx
        cmpq    %rcx, %rax
        jne     L155
        movq    %rax, 8(%rsp)
        movq    %rsi, (%rsp)
        movabsq $"__avx__!", %rax
        movabsq $139712568057360, %rdi  # imm = 0x7F115DFD5E10
        movabsq $139712568057456, %rsi  # imm = 0x7F115DFD5E70
        movabsq $139712568058800, %rdx  # imm = 0x7F115DFD63B0
        movabsq $139712568058992, %rcx  # imm = 0x7F115DFD6470
        leaq    8(%rsp), %r8
        vxorpd  %xmm0, %xmm0, %xmm0
        callq   *%rax
        vextractf64x4   $1, %zmm0, %ymm1
        vaddpd  %zmm1, %zmm0, %zmm0
        vextractf128    $1, %ymm0, %xmm1
        vaddpd  %xmm1, %xmm0, %xmm0
        vpermilpd       $1, %xmm0, %xmm1 # xmm1 = xmm0[1,0]
        vaddsd  %xmm1, %xmm0, %xmm0
        addq    $40, %rsp
        vzeroupper
        retq
L155:
        movabsq $jl_system_image_data, %rax
        movq    %rax, 16(%rsp)
        movq    %r9, 24(%rsp)
        movq    %rsi, 32(%rsp)
        movabsq $jl_invoke, %rax
        movabsq $jl_system_image_data, %rdi
        leaq    16(%rsp), %rsi
        movabsq $139712388188304, %rcx  # imm = 0x7F115344C890
        movl    $3, %edx
        callq   *%rax
        ud2
```
while the default version is inlined:
```asm
# julia> @code_native debuginfo=:none mydotavx(A, B)
        .text
        subq    $24, %rsp
        movq    8(%rdi), %rax
        movq    %rax, %rcx
        sarq    $63, %rcx
        andnq   %rax, %rcx, %r8
        movq    8(%rsi), %rax
        movq    %rax, %rcx
        sarq    $63, %rcx
        andnq   %rax, %rcx, %rax
        cmpq    %rax, %r8
        jne     L622
        movq    (%rdi), %rdx
        movq    (%rsi), %rax
        leal    -1(%r8), %ecx
        andl    $7, %ecx
        movb    $-2, %sil
        shlb    %cl, %sil
        cmpq    $7, %rcx
        notb    %sil
        movzbl  %sil, %esi
        movl    $255, %r9d
        cmovnel %esi, %r9d
        leaq    -30(%r8), %rdi
        cmpq    $2, %rdi
        jl      L266
        vxorpd  %xmm0, %xmm0, %xmm0
        movl    $24, %ecx
        vxorpd  %xmm2, %xmm2, %xmm2
        vxorpd  %xmm3, %xmm3, %xmm3
        vxorpd  %xmm1, %xmm1, %xmm1
        nopl    (%rax,%rax)
L128:
        movq    %rcx, %rsi
        vmovupd -192(%rdx,%rcx,8), %zmm4
        vmovupd -128(%rdx,%rcx,8), %zmm5
        vmovupd -64(%rdx,%rcx,8), %zmm6
        vmovupd (%rdx,%rcx,8), %zmm7
        vfmadd231pd     -192(%rax,%rcx,8), %zmm4, %zmm1 # zmm1 = (zmm4 * mem) + zmm1
        vfmadd231pd     -128(%rax,%rcx,8), %zmm5, %zmm3 # zmm3 = (zmm5 * mem) + zmm3
        vfmadd231pd     -64(%rax,%rcx,8), %zmm6, %zmm2 # zmm2 = (zmm6 * mem) + zmm2
        vfmadd231pd     (%rax,%rcx,8), %zmm7, %zmm0 # zmm0 = (zmm7 * mem) + zmm0
        addq    $32, %rcx
        addq    $9, %rsi
        cmpq    %rdi, %rsi
        jl      L128
        addq    $-23, %rcx
        cmpq    %rcx, %r8
        jl      L563
L219:
        leaq    -8(%r8), %rsi
        cmpq    %rcx, %rsi
        jge     L298
        kmovd   %r9d, %k1
        vmovupd -8(%rdx,%rcx,8), %zmm4 {%k1} {z}
        vmovupd -8(%rax,%rcx,8), %zmm5 {%k1} {z}
        vfmadd231pd     %zmm5, %zmm4, %zmm1 {%k1} # zmm1 = (zmm4 * zmm5) + zmm1
        jmp     L563
L266:
        movl    $1, %ecx
        vxorpd  %xmm1, %xmm1, %xmm1
        vxorpd  %xmm3, %xmm3, %xmm3
        vxorpd  %xmm2, %xmm2, %xmm2
        vxorpd  %xmm0, %xmm0, %xmm0
        cmpq    %rcx, %r8
        jl      L563
        jmp     L219
L298:
        leaq    -16(%r8), %rsi
        cmpq    %rcx, %rsi
        jge     L367
        vmovupd -8(%rdx,%rcx,8), %zmm4
        kmovd   %r9d, %k1
        vmovupd 56(%rdx,%rcx,8), %zmm5 {%k1} {z}
        vmovupd 56(%rax,%rcx,8), %zmm6 {%k1} {z}
        vfmadd231pd     -8(%rax,%rcx,8), %zmm4, %zmm1 # zmm1 = (zmm4 * mem) + zmm1
        vfmadd231pd     %zmm6, %zmm5, %zmm3 {%k1} # zmm3 = (zmm5 * zmm6) + zmm3
        jmp     L563
L367:
        leaq    -24(%r8), %rsi
        cmpq    %rcx, %rsi
        jge     L455
        vmovupd -8(%rdx,%rcx,8), %zmm4
        vmovupd 56(%rdx,%rcx,8), %zmm5
        kmovd   %r9d, %k1
        vmovupd 120(%rdx,%rcx,8), %zmm6 {%k1} {z}
        vmovupd 120(%rax,%rcx,8), %zmm7 {%k1} {z}
        vfmadd231pd     -8(%rax,%rcx,8), %zmm4, %zmm1 # zmm1 = (zmm4 * mem) + zmm1
        vfmadd231pd     56(%rax,%rcx,8), %zmm5, %zmm3 # zmm3 = (zmm5 * mem) + zmm3
        vfmadd231pd     %zmm7, %zmm6, %zmm2 {%k1} # zmm2 = (zmm6 * zmm7) + zmm2
        jmp     L563
L455:
        addq    $-32, %r8
        cmpq    %rcx, %r8
        jge     L563
        vmovupd -8(%rdx,%rcx,8), %zmm4
        vmovupd 56(%rdx,%rcx,8), %zmm5
        vmovupd 120(%rdx,%rcx,8), %zmm6
        kmovd   %r9d, %k1
        vmovupd 184(%rdx,%rcx,8), %zmm7 {%k1} {z}
        vmovupd 184(%rax,%rcx,8), %zmm8 {%k1} {z}
        vfmadd231pd     -8(%rax,%rcx,8), %zmm4, %zmm1 # zmm1 = (zmm4 * mem) + zmm1
        vfmadd231pd     56(%rax,%rcx,8), %zmm5, %zmm3 # zmm3 = (zmm5 * mem) + zmm3
        vfmadd231pd     120(%rax,%rcx,8), %zmm6, %zmm2 # zmm2 = (zmm6 * mem) + zmm2
        vfmadd231pd     %zmm8, %zmm7, %zmm0 {%k1} # zmm0 = (zmm7 * zmm8) + zmm0
L563:
        vaddpd  %zmm0, %zmm3, %zmm0
        vaddpd  %zmm2, %zmm1, %zmm1
        vaddpd  %zmm0, %zmm1, %zmm0
        vextractf64x4   $1, %zmm0, %ymm1
        vaddpd  %zmm1, %zmm0, %zmm0
        vextractf128    $1, %ymm0, %xmm1
        vaddpd  %xmm1, %xmm0, %xmm0
        vpermilpd       $1, %xmm0, %xmm1 # xmm1 = xmm0[1,0]
        vaddsd  %xmm1, %xmm0, %xmm0
        addq    $24, %rsp
        vzeroupper
        retq
L622:
        movabsq $jl_system_image_data, %rax
        movq    %rax, (%rsp)
        movq    %rdi, 8(%rsp)
        movq    %rsi, 16(%rsp)
        movabsq $jl_invoke, %rax
        movabsq $jl_system_image_data, %rdi
        movq    %rsp, %rsi
        movabsq $139712388188304, %rcx  # imm = 0x7F115344C890
        movl    $3, %edx
        callq   *%rax
        ud2
```

For more expensive functions, the `inline=false` version can get better performance:
```julia
julia> using LoopVectorization, BenchmarkTools

julia> M = K = N = 128
128

julia> function AmulBavx1!(C, A, B)
           @avx for m ∈ 1:size(A,1), n ∈ 1:size(B,2)
               Cₘₙ = zero(eltype(C))
               for k ∈ 1:size(A,2)
                   Cₘₙ += A[m,k] * B[k,n]
               end
               C[m,n] = Cₘₙ
           end
       end
AmulBavx1! (generic function with 1 method)

julia> function AmulBavx1_noinline!(C, A, B)
           @avx inline=false for m ∈ 1:size(A,1), n ∈ 1:size(B,2)
               Cₘₙ = zero(eltype(C))
               for k ∈ 1:size(A,2)
                   Cₘₙ += A[m,k] * B[k,n]
               end
               C[m,n] = Cₘₙ
           end
       end
AmulBavx1_noinline! (generic function with 1 method)

julia> A = rand(M, K); B = rand(K, N); C1 = Matrix{Float64}(undef, M, N); C2 = similar(C1);

julia> @benchmark AmulBavx1!($C2, $A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     37.724 μs (0.00% GC)
  median time:      37.882 μs (0.00% GC)
  mean time:        37.956 μs (0.00% GC)
  maximum time:     109.022 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark AmulBavx1_noinline!($C1, $A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     36.693 μs (0.00% GC)
  median time:      36.893 μs (0.00% GC)
  mean time:        36.951 μs (0.00% GC)
  maximum time:     120.853 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> C1 ≈ C2 ≈ A * B
true

julia> @code_native debuginfo=:none AmulBavx1_noinline!(C1, A, B) # not inlined
        .text
        pushq   %rbp
        movq    %rsp, %rbp
        pushq   %rbx
        andq    $-32, %rsp
        subq    $96, %rsp
        vxorps  %xmm0, %xmm0, %xmm0
        vmovaps %ymm0, (%rsp)
        movq    $0, 32(%rsp)
        movq    %rsi, 64(%rsp)
        movq    %fs:0, %rbx
        movq    $12, (%rsp)
        movq    -15712(%rbx), %rax
        movq    %rax, 8(%rsp)
        movq    %rsp, %rax
        movq    %rax, -15712(%rbx)
        movq    (%rsi), %r10
        movq    8(%rsi), %r9
        movq    16(%rsi), %rax
        movq    32(%rax), %rcx
        movq    24(%r9), %rdx
        movq    32(%r9), %rsi
        movq    %rdx, 40(%rsp)
        movq    %rcx, 48(%rsp)
        movq    %rsi, 56(%rsp)
        movq    %r9, 32(%rsp)
        movq    %rax, 24(%rsp)
        movq    %r10, 16(%rsp)
        movabsq $"__avx__!", %r11
        movabsq $139712396595344, %rdi  # imm = 0x7F1153C51090
        movabsq $139712396595440, %rsi  # imm = 0x7F1153C510F0
        movabsq $139712396597840, %rdx  # imm = 0x7F1153C51A50
        movabsq $139712396597936, %rcx  # imm = 0x7F1153C51AB0
        leaq    40(%rsp), %r8
        pushq   %r10
        pushq   %rax
        vzeroupper
        callq   *%r11
        addq    $16, %rsp
        movq    8(%rsp), %rax
        movq    %rax, -15712(%rbx)
        movabsq $jl_system_image_data, %rax
        leaq    -8(%rbp), %rsp
        popq    %rbx
        popq    %rbp
        retq
```

The cost of a function call is negligible on the order of microseconds.

You've done a lot of work with Julia's inliner, so you know better than I do what the inline cost modeling should look like. Any thoughts or suggestions?
I think using LoopVectorization's cost modeling to decide would be the best default (better than `inline = true`, anyway).

Reply:
Aha! I was expecting a difference in the code produced by `_avx_!`, so I stashed the outcome of its `avx_body` call in a `Ref` and then extracted it after making the call. I hadn't realized I was looking at it in the wrong way. The difference is quite clear with `@code_typed`.

Sorry, I should have figured that out, but I was so focused on trying to understand the two `setup` calls that I failed to notice that a key difference was in `generate_call`.

> You've done a lot of work with Julia's inliner, so you know better than I do what the inline cost modeling should look like. Any thoughts or suggestions?

The reality is that while the new inlining engine is vastly smarter than the old one (and generally makes better decisions), it's still often wrong. IMO the big problem is a failure to account for optimizations that can occur only when inlining; for example, if a block of code creates a view and then passes that view to a helper function, inlining the helper might allow Julia to elide the view-creation itself.  But that's not accounted for now. (It wouldn't be phenomenally difficult to implement, but I think there's every reason to expect that it would make our current compile latencies look like an eyeblink, and that seems like too high of a price to pay.)

I've pushed a new version and modified the PR title to indicate that I think this might be ready to merge.